iptables/src/ip6tables_ext.cpp

```diff
@@ -0,0 +1,209 @@
+
+  if (!s.ok()) {
+    TLOG << "Error fetching matches from ip6tables-save: " << s.toString();
+  }
+
+  for (const auto& table : getIp6tablesNames()) {
+    const auto& matches = match_map.find(table);
+
+    if (matches == match_map.end()) {
+      TLOG << "Couldn't associate table " << table << " with a list of matches";
+      return results;
+    }
+
+    s = genIptablesRules(table, matches->second, results);
+
+    if (!s.ok()) {
+      TLOG << "Error while fetching table rules: " << s.toString();
+      return results;
+    }
+  }
+
+  return results;
+}
+
+osquery::Status Ip6tablesExtTable::genIptablesRules(
+    const std::string& filter,
+    const MatchChain& matches,
+    osquery::QueryData& results) {
+  // Initialize the access to ip6tc
+  auto handle = ip6tc_init(filter.c_str());
+  if (handle == nullptr) {
+    return osquery::Status(1, "Couldn't initialize ip6tables handle");
+  }
+
+  // Iterate through chains
+  for (auto chain = ip6tc_first_chain(handle); chain != nullptr;
+       chain = ip6tc_next_chain(handle)) {
+    // NOTE(ww): Rules are 1-based in libip6tc, as evidenced by
+    // ip6tc_read_counter.
+    unsigned long ruleno = 1;
+    const auto& match_pair = matches.find(chain);
+
+    if (match_pair == matches.end()) {
+      TLOG << "couldn't associate " << chain << " with a list of matches";
+      return osquery::Status(1, "couldn't associate chain with a match list!");
+    }
+
+    const auto& match_list = match_pair->second;
+
+    // Iterating through all the rules per chain
+    for (auto chain_rule = ip6tc_first_rule(chain, handle);
+         chain_rule != nullptr;
+         chain_rule = ip6tc_next_rule(chain_rule, handle)) {
+      osquery::Row r;
+
+      r["table_name"] = TEXT(filter);
+      r["chain"] = TEXT(chain);
+      r["packets"] = BIGINT(chain_rule->counters.pcnt);
+      r["bytes"] = BIGINT(chain_rule->counters.bcnt);
+      r["ruleno"] = INTEGER(ruleno);
+
+      if (ruleno - 1 < match_list.size()) {
+        const auto& match_entry = match_list.at(ruleno - 1);
+
+        r["match"] = TEXT(std::get<0>(match_entry));
+        r["target"] = TEXT(std::get<1>(match_entry));
+        r["target_options"] = TEXT(std::get<2>(match_entry));
+      } else {
+        TLOG << "rule number mismatch: " << (ruleno - 1)
+             << " >= " << match_list.size();
+        r["match"] = TEXT("");
+        r["target"] = TEXT("");
+        r["target_options"] = TEXT("");
+      }
+
+      if (chain_rule->target_offset) {
+        // This is basically the IP6T_MATCH_ITERATE macro from ip6tables,
+        // but without the GNU C magic (void pointer arithmetic,
+        // macro expression extensions).
+        const xt_entry_match* match;
+        for (int i = sizeof(ip6t_entry); i < chain_rule->target_offset;
+             i += match->u.match_size) {
+          match = reinterpret_cast<const xt_entry_match*>(
+              reinterpret_cast<const char*>(chain_rule) + i);
```

Comment:
super nit, it occurs to me that maybe we could be super fancy here and rather than casting to a `const char *`, we could use something like `std::uintptr_t` which as i understand the proper type for doing arbitrary integer arithmetic on a pointer. i tried, and it also requires changing `i` to an `unsigned int` due to some compiler error about implicit type conversion

this is something so small as to barely even matter, but just though i'd point it out -- feel free to disregard!

Reply:
I've run into this before! `uintptr_t`'s semantics are actually more limited than `char *`'s -- the C spec guarantees that casting a pointer to a `uintptr_t` and then casting it back will yield the same value, but does *not* guarantee that casting a modified `uintptr_t` back into a pointer will be valid (since the layout of `uintptr_t` is unspecified).

Some links: https://stackoverflow.com/questions/43418618/c-why-cast-to-uintptr-t-vs-char-when-doing-pointer-arithmetic and https://stackoverflow.com/questions/19473107/if-i-want-to-do-regular-arithmetic-not-pointer-arithmetic-on-a-pointer-should/19476805#19476805

Reply:
TIL, thanks!